### PR TITLE
postcss-nesting : less :is pseudo insertion

### PR DESCRIPTION
**Pull request by romainmenke**
## postcss-nesting 

- [x] aware of selector specificity and meaning -> less `:is`
- [x] able to sort compound selectors -> less `:is`
- [x] able to expand nested selector lists -> less `:is`
- [x] option to disable `:is` completely
- [x] document code (_please ask for more comments if anyone feels it's needed_)
- [x] document option
- [x] update CLI options

## postcss-preset-env

- [x] plugin specific options based on browsers provided as options
- [x] automatically disables `:is` in `postcss-nesting` when `:is` is not supported 
- [x] added tests for the switching between with or without `:is` pseudo's 

-----

Also compared this with an [older version](https://github.com/csstools/postcss-nesting/releases/tag/9.0.0) of `postcss-nesting` :

_results are for plugin with `:is` allowed_

```diff
diff --git a/plugins/postcss-nesting/test/basic.expect.css b/plugins/postcss-nesting/test/basic.expect.css
index 759fdf04..d2780a9b 100644
--- a/plugins/postcss-nesting/test/basic.expect.css
+++ b/plugins/postcss-nesting/test/basic.expect.css
@@ -39,7 +39,7 @@ a {
 body a {
                order: 11
 }
-html body a {
+html :is(body a) {
                        order: 12
 }
 body a {
diff --git a/plugins/postcss-nesting/test/complex.expect.css b/plugins/postcss-nesting/test/complex.expect.css
index 6557492e..2cdee7d5 100644
--- a/plugins/postcss-nesting/test/complex.expect.css
+++ b/plugins/postcss-nesting/test/complex.expect.css
@@ -1,7 +1,6 @@
 body > p, body > ul {
-       margin: 0;
-
-       & ~ & {
+       margin: 0
+}
+:is(body > p) ~ :is(body > p), :is(body > p) ~ :is(body > ul), :is(body > ul) ~ :is(body > p), :is(body > ul) ~ :is(body > ul) {
                margin-top: 0;
        }
-}
diff --git a/plugins/postcss-nesting/test/direct.expect.css b/plugins/postcss-nesting/test/direct.expect.css
index fad601f5..ec8fc6cc 100644
--- a/plugins/postcss-nesting/test/direct.expect.css
+++ b/plugins/postcss-nesting/test/direct.expect.css
@@ -1,13 +1,13 @@
 a, b {
        order: 1;
 }
-a c, a d, b c, b d {
+a c, b c, a d, b d {
                order: 2;
        }
-a c e, a c f, a d e, a d f, b c e, b c f, b d e, b d f {
+a c e, b c e, a d e, b d e, a c f, b c f, a d f, b d f {
                        order: 3;
                }
-a c, a d, b c, b d {
+a c, b c, a d, b d {
                order: 4;
        }
 a, b {
@@ -16,13 +16,13 @@ a, b {
 a, b {
        order: 1;
 }
-a c, a d, b c, b d {
+a c, b c, a d, b d {
                order: 2;
 }
-a c e, a c f, a d e, a d f, b c e, b c f, b d e, b d f {
+a c e, b c e, a d e, b d e, a c f, b c f, a d f, b d f {
                        order: 3;
 }
-a c, a d, b c, b d {
+a c, b c, a d, b d {
                order: 4;
 }
 a, b {
diff --git a/plugins/postcss-nesting/test/ignore.expect.css b/plugins/postcss-nesting/test/ignore.expect.css
index 133a2e70..13bdd8e4 100644
--- a/plugins/postcss-nesting/test/ignore.expect.css
+++ b/plugins/postcss-nesting/test/ignore.expect.css
@@ -7,11 +7,9 @@ a, b {
 & e {
        order: 3;
 }
-f {
-       &g {
+f:is(g) {
                order: 5;
        }
-}
 a, b {
        order: 1;
        @nest c, d {
@@ -21,10 +19,10 @@ a, b {
 @nest & e {
        order: 3;
 }
+f:is(h) {
+               order: 5
+}
 f {
-       @nest &h {
-               order: 5;
-       }
        @nest &|i {
                order: 6;
        }

```

### CODE REVIEW

**Comment by romainmenke** [comment]
Input :

```css
.error,
#404 {
	&:hover>.baz {
		color: red;
	}
}
```

Before this PR :

```css
:is(.error,#404):hover>.baz {
		color: red;
	}
```

--------

After this PR :

```css
:is(.error,#404):hover>.baz {
		color: red;
	}
```

With `:is()` disabled

```css
.error:hover>.baz, #404:hover>.baz {
		color: red;
	}
```

Without `:is()` the selector would have different specificity than required by spec, but it still works in browsers without `:is()` support.

`:is()` is disabled via `preset-env` if any browser listed does not support it.

File: plugin-packs/postcss-preset-env/.tape.js [hunk 1]
```diff
@@ -20,6 +20,32 @@ module.exports = {
...
 		}
 	},
+	'basic:ch88-ff78': {
+		message: 'supports { browsers: "chrome >= 88, firefox >= 78", stage: 0 } usage',
+		options: {
+			browsers: 'chrome >= 88, firefox >= 78',
+			stage: 0
+		}
+	},
+		'basic:ch88-ff78:no-is-pseudo': {
```
Comment (Antonio-Laguna) [inline plugin-packs/postcss-preset-env/.tape.js:30]: [nit] Odd indentation

File: plugin-packs/postcss-preset-env/.tape.js [hunk 2]
```diff
@@ -20,6 +20,32 @@ module.exports = {
 			browsers: 'chrome >= 38'
 		}
 	},
+	'basic:ch88-ff78': {
+		message: 'supports { browsers: "chrome >= 88, firefox >= 78", stage: 0 } usage',
```
Comment (Antonio-Laguna) [inline plugin-packs/postcss-preset-env/.tape.js:24]: Thoughts on updating this message and below to be more descriptive of what we intend? Something within these lines: 

`Modern browsers should get :is`
`Is can be disabled`,

etc

Feel like when reading these test cases it's not clear the purpose they serve.

File: plugins/postcss-nesting/README.md [hunk 3]
```diff
@@ -71,6 +71,73 @@ import postcssNesting from "https://cdn.jsdelivr.net/npm/postcss-nesting@10/mod.
...
+
+```css
+.alpha > .beta + .alpha > .beta {
+	order: 2;
+}
+```
+
+_this is a different selector than expected_
+_avoid these cases when you disable `:is()`_
+_writing the selector without nesting is advised here_
```
Comment (Antonio-Laguna) [inline plugins/postcss-nesting/README.md:139]: I think we could do some more teaching here if we want to. This is discouraging but not actually clear, look at this

https://polypane.app/css-specificity-calculator/#selector=.alpha%20%3E%20.beta%20%2B%20.alpha%20%3E%20.beta

vs 

https://polypane.app/css-specificity-calculator/#selector=%3Ais(.alpha%20%3E%20.beta)%20%2B%20%3Ais(.alpha%20%3E%20.beta)

Which is the same (shouldn't?) 

Let me know if you want to discuss this further.

File: plugins/postcss-nesting/src/lib/merge-selectors/merge-selectors.js [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+				// If the from selector is simple we extract the first non root, non selector node
+				if (fromSelectorAST.type === 'root' && fromSelectorAST.nodes.length === 1) {
+					fromSelectorAST = fromSelectorAST.nodes[0];
+				}
+
+				const fromSelectorWithIsAST = parser().astSync(`:is(${fromSelectorAST.toString()})`);
+
+				const fromIsSimple = isSimpleSelector(fromSelectorAST.nodes[0]); // this function looks at the parent of the node passed as an argument
+				const fromIsCompound = isCompoundSelector(fromSelectorAST.nodes[0]); // this function looks at the parent of the node passed as an argument
```
Comment (Antonio-Laguna) [inline plugins/postcss-nesting/src/lib/merge-selectors/merge-selectors.js:67]: ❤️  all of these comments

**Review by Antonio-Laguna — COMMENTED** [review]
@romainmenke this is awesome! Thanks for all of the extra comments and for this effort, looks terrific, and can't wait to get it out.

I've left some minor comments to increase clarity mostly in some areas. Open for discussion any time 😄

File: plugins/postcss-nesting/README.md (hunk 3 above)
Comment (romainmenke) [reply plugins/postcss-nesting/README.md]: I like that!
I added links to this site where applicable and I also split up the edge cases.

Mixing them was indeed confusing :

https://github.com/csstools/postcss-plugins/blob/postcss-nesting-less-is-pseudo--rational-curly-coated-retriever-dedd642a50/plugins/postcss-nesting/README.md#options

File: plugin-packs/postcss-preset-env/.tape.js (hunk 2 above)
Comment (romainmenke) [reply plugin-packs/postcss-preset-env/.tape.js]: Updated to : 

```js
	'basic:ch88-ff78': {
		message: 'uses :is pseudo for nesting with modern browsers { browsers: "chrome >= 88, firefox >= 78", stage: 0 }',
		options: {
			browsers: 'chrome >= 88, firefox >= 78',
			stage: 0
		}
	},
	'basic:ch88-ff78:no-is-pseudo': {
		message: ':is pseudo for nesting can be disable with modern browsers  { browsers: "chrome >= 88, firefox >= 78", stage: 0, features: { nesting-rules: { noIsPseudoSelector: true } } } usage',
		options: {
			browsers: 'chrome >= 88, firefox >= 78',
			stage: 0,
			features: {
				'nesting-rules': {
					noIsPseudoSelector: true
				}
			}
		}
	},
	'basic:ch88-ff78-saf10': {
		message: 'does not use :is pseudo for nesting with an older browser { browsers: "chrome >= 88, firefox >= 78, safari >= 10", stage: 0 } usage',
		options: {
			browsers: 'chrome >= 88, firefox >= 78, safari >= 10',
			stage: 0
		}
	},
```

File: plugin-packs/postcss-preset-env/.tape.js (hunk 1 above)
Comment (romainmenke) [reply plugin-packs/postcss-preset-env/.tape.js]: fixed

**Review by Antonio-Laguna — APPROVED** [review]
Nice work!

**Review by jonathantneal — APPROVED** [review]
Well done.

**Comment by romainmenke** [comment]
Thank you for reviewing this 🙇

**Pull request merged**